returnn/datasets/basic.py

```diff
@@ -111,6 +111,8 @@ def __init__(
         min_chunk_size=0,
         chunking_variance=0,
         estimated_num_seqs=None,
+        num_data_shards=1,
+        data_shard_index=0,
```

Comment:
As discussed, should be:
```suggestion
        _num_data_shards=1,
        _data_shard_index=0,
```

Reply:
> They cannot have a leading underscore, because the `Dataset.reduce` logic will treat them special if they do.

I don't exactly understand this argument. We can just fix `Dataset.reduce` if there is some problem? But also, do you mean `Dataset.__reduce__`? And how does it treat them special?

`_dataset_extend_default_kwargs_from_parent_dataset` can surely read internal things from the `Dataset`, that's not really a problem. (We should not blindly follow the rules by the linting, but think about what makes sense.)

In any case, they really should also be private attributes (I forgot to comment that), not public.